controller-server/src/test/java/com/yahoo/vespa/hosted/controller/deployment/DeploymentContext.java

```diff
@@ -0,0 +1,496 @@
+import static org.junit.Assert.assertFalse;
+import static org.junit.Assert.assertNotNull;
+import static org.junit.Assert.assertTrue;
+
+/**
+ * A deployment context for an application. This allows fine-grained control of the deployment of an application's
+ * instances.
+ *
+ * References to this should be acquired through {@link InternalDeploymentTester#deploymentContext}.
+ *
+ * Tester code that is not specific to deployments should be added to either {@link ControllerTester} or
+ * {@link InternalDeploymentTester} instead of this class.
+ *
+ * @author mpolden
+ */
+public class DeploymentContext {
+
+    // Set a long interval so that maintainers never do scheduled runs during tests
+    private static final Duration maintenanceInterval = Duration.ofDays(1);
+
+    public static final ApplicationPackage applicationPackage = new ApplicationPackageBuilder()
+            .athenzIdentity(AthenzDomain.from("domain"), AthenzService.from("service"))
+            .upgradePolicy("default")
+            .region("us-central-1")
+            .parallel("us-west-1", "us-east-3")
+            .emailRole("author")
+            .emailAddress("b@a")
+            .build();
+
+    public static final ApplicationPackage publicCdApplicationPackage = new ApplicationPackageBuilder()
+            .athenzIdentity(AthenzDomain.from("domain"), AthenzService.from("service"))
+            .upgradePolicy("default")
+            .region("aws-us-east-1c")
+            .emailRole("author")
+            .emailAddress("b@a")
+            .trust(generateCertificate())
+            .build();
+
+    private final TenantAndApplicationId applicationId;
+    private final ApplicationId instanceId;
+    private final TesterId testerId;
+    private final JobController jobs;
+    private final RoutingGeneratorMock routing;
+    private final MockTesterCloud cloud;
+    private final JobRunner runner;
+    private final ControllerTester tester;
+    private final ReadyJobsTrigger readyJobsTrigger;
+    private final NameServiceDispatcher nameServiceDispatcher;
+
+    private ApplicationVersion lastSubmission = null;
+    private boolean deferDnsUpdates = false;
+
+    public DeploymentContext(ApplicationId instanceId, ControllerTester tester, JobRunner runner) {
+        var jobControl = new JobControl(tester.controller().curator());
+        this.applicationId = TenantAndApplicationId.from(instanceId);
+        this.instanceId = instanceId;
+        this.testerId = TesterId.of(instanceId);
+        this.jobs = tester.controller().jobController();
+        this.routing = tester.serviceRegistry().routingGeneratorMock();
+        this.cloud = tester.serviceRegistry().testerCloud();
+        this.runner = runner;
+        this.tester = tester;
+        this.readyJobsTrigger = new ReadyJobsTrigger(tester.controller(), maintenanceInterval, jobControl);
+        this.nameServiceDispatcher = new NameServiceDispatcher(tester.controller(), maintenanceInterval, jobControl,
```

Comment:
Things that can be safely shared across contexts should probably continue living in `InternalDeploymentTester`, so these maintainers can be moved back or perhaps to `ControllerTester`.

Reply:
Yes.